### ISSUE TITLE
Unfortunate optimization with Option<bool> vs Option<uint>

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#20149]
Currently, given this program:

```rust
#[inline(never)]
fn get_bool(slot: &mut Option<bool>) -> &mut bool {
    if slot.is_none() {
        *slot = Some(true);
    }
    slot.as_mut().unwrap()
}

#[inline(never)]
fn get_uint(slot: &mut Option<uint>) -> &mut uint {
    if slot.is_none() {
        *slot = Some(1);
    }
    slot.as_mut().unwrap()
}

fn main() {
    println!("{}", get_bool(&mut None));
    println!("{}", get_uint(&mut None));
}
```

It generates [this IR](https://gist.github.com/alexcrichton/ecf9578b16849a12bd7b).

The "unfortunate" part here is that a [call to failure](https://gist.github.com/alexcrichton/ecf9578b16849a12bd7b#file-foo-ll-L49-L51) is generated [for the `get_bool`](https://gist.github.com/alexcrichton/ecf9578b16849a12bd7b#file-foo-ll-L28-L55) function where no failure is generated for the [`get_uint` function](https://gist.github.com/alexcrichton/ecf9578b16849a12bd7b#file-foo-ll-L67-L87).

There's a lot going on here so I'm not sure if it's our codegen or LLVM's codegen, but I would personally expect both cases to optimize down to never unwinding because it's known that neither `Option` can ever be `None`.

**Comment by dotdash** [comment]
http://reviews.llvm.org/D6829 -- This teaches LLVM to handle the aggregate and extract the loaded value from it, allowing it to optimize the code even without #20161. It also handles the example I presented there where the deferred load breaks an optimization.

**Comment by alexcrichton** [comment]
Wow, thanks so much for looking into this @dotdash!

**Comment by huonw** [comment]
Closed by #20755.
